vent/sqldb/sqldb.go

```
@@ -39,132 +41,198 @@ const (
-	insValues := ""
-	updValues := ""
+	var tables []Table
+	tables = append(tables, table)
+	return db.UpsertRows(tables)
+	/*
+		query, pointers := getUpsertQuery(db.schema, table)
 
-	cols := 0
-	nkeys := 0
-	for _, tableColumn := range table.Columns {
-		cols++
-		if !tableColumn.Primary {
-			nkeys++
+		db.log.Debug("msg", "Upsert row", "query", clean(query), "value", pointers)
+		_, err := db.DB.Exec(query, pointers...)
+
+		if pqErr, ok := err.(*pq.Error); ok {
+			if pqErr.Code == errUndefinedTable {
+				//create table and retry
+				db.log.Warn("msg", "Table not found", "value", table.TableName)
+				ctErr := db.createTable(table)
+				if ctErr != nil {
+					return ctErr
+				}
+			} else if pqErr.Code == errUndefinedColumn {
+				//alter table and retry
+				db.log.Warn("msg", "Column not found")
+				alErr := db.alterTable(table)
+				if alErr != nil {
+					return alErr
+				}
+			} else {
+				db.log.Error("msg", "Error upserting row", "err", pqErr)
+				return pqErr
+			}
+
+			//retry
+			db.log.Debug("msg", "Upsert row", "query", clean(query), "value", pointers)
+			_, err = db.DB.Exec(query, pointers...)
+			return err
+		}
+		return err
+	*/
+
+}
+
+// SynchronizeDB  against PGSQL
+func (db *SQLDB) SynchronizeDB(tables []Table) error {
+	for _, table := range tables {
+		found, err := db.FindTable(table.TableName)
+		if err != nil {
+			return err
+		}
+
+		if found {
+			err = db.alterTable(table)
+		} else {
+			err = db.createTable(table)
+		}
+		if err != nil {
+			return err
 		}
 	}
+	return nil
+}
 
-	pointers := make([]interface{}, cols+nkeys)
-	containers := make([]string, cols+nkeys)
+// UpsertRows inserts or updates a multiple rows in PGSQL and writes _bosbarmot_log
+func (db *SQLDB) UpsertRows(tables []Table) error {
+	tx, err := db.DB.Begin()
 
-	// For Each column in Table
-	p := 0
-	cnk := 0
-	for colName, tableColumn := range table.Columns {
+	if err != nil {
+		db.log.Error("msg", "Error beginning transaction", "err", err)
```

You can probably just return the error here without logging anything, because the caller will probably log it too.